payments/models.py

```diff
@@ -154,3 +169,356 @@ def __str__(self):
+
+        self.product = product
+
+    def _assign_lease_from_object_id(self):
+        # Try to get a BerthLease (BL)
+        lease = BerthLease.objects.filter(id=self._lease_object_id).first()
+        # If the BL was not found, try getting a WinterStorageLease
+        lease = (
+            lease
+            if lease
+            else WinterStorageLease.objects.filter(id=self._lease_object_id).first()
+        )
+        # If for some reason neither was found, raise an error
+        if not lease:
+            raise ValidationError(_("The lease passed is not valid"))
+
+        self.lease = lease
+
+    def save(self, *args, **kwargs):  # noqa: C901  is too complex
+        self.full_clean()
+
+        # If the product is being added from the admin (only the ID is passed)
+        if self._product_object_id and not self.product:
+            self._assign_product_from_object_id()
+
+        # If the lease is being added from the admin (only the ID is passed)
+        if self._lease_object_id and not self.lease:
+            self._assign_lease_from_object_id()
+
+        creating = self._state.adding
+        # If the product instance is being passed
+        # Price has to be assigned before saving but only if creating
+        if creating and self.product:
+            # Assign the price and tax from the product
+            if not self.price:
+                self.price = self.product.price_value
+            if not self.tax_percentage:
+                self.tax_percentage = self.product.tax_percentage
+
+        old_instance = Order.objects.filter(id=self.id).first()
+        # Save before adding additional products to have access to self
+        super().save(*args, **kwargs)
+
+        # Create OrderLines for the corresponding services only:
+        #   - If it's creating the order
+        #   - If the previous lease value was None (setting a lease)
+        # For now, it only applies BerthLeases, since WinterStorageLeases include the
+        # services on the m2 price.
+        if (creating or (old_instance and not old_instance.lease)) and isinstance(
+            self.lease, BerthLease
+        ):
+            self._create_order_lines(self.lease.berth.pier)
+
+
+class OrderLine(UUIDModel, TimeStampedModel):
+    order = models.ForeignKey(
+        Order,
+        verbose_name=_("order"),
+        related_name="order_lines",
+        on_delete=models.CASCADE,
+    )
+    product = models.ForeignKey(
+        AdditionalProduct,
+        verbose_name=_("product"),
+        related_name="orders",
+        on_delete=models.CASCADE,
+    )
+    quantity = models.PositiveSmallIntegerField(
+        verbose_name=_("quantity"), default=1, validators=[MinValueValidator(1)],
+    )
+    price = models.DecimalField(
+        verbose_name=_("price"),
+        max_digits=10,
+        decimal_places=2,
+        validators=[MinValueValidator(Decimal("0.01"))],
+        blank=True,
+        null=True,
```

Comment:
but then the `price` (and `tax_percentage`) down below would not be nullable, maybe? they would store the price and tax of the product at the time of order line creation...

Reply:
The only reason atm to have them blankable/nullable would be to allow creating `OrderLine`s from the admin without explicitly passing them (when passing a product).

Reply:
hm, let's have them a blankable then, maybe? since that would allow leaving them empty in Django admin, but would block not setting any value

Reply:
i.e. without `null=True`

Reply:
Great idea! 👍 I'll do the same for price/tax on `Order` and `OrderLine`

Reply:
yeah, I was on the edge about allowing `null`s for prices and taxes on the `Order`. this might lead to orders with empty prices and distort the bookkeeping, better to ensure the price is always set on the DB level